File: app/src/at-phuongle/java/com/asiantech/intern20summer1/intro/RainbowOfClarity.java
package com.asiantech.intern20summer1.intro;

import java.util.ArrayList;
import java.util.List;

public class RainbowOfClarity {
    /**
     * Write a function that checks if the given character is a digit or not.
     */
    boolean isDigit(char symbol) {
        return Character.isDigit(symbol);
    }

    /**
     * Write a function that returns its encoding.
     */
    String lineEncoding(String s) {
        s += "#";
        int count = 1;
        String result = "";

        for (int i = 1; i < s.length(); i++) {
            if (s.charAt(i) == s.charAt(i - 1)) {
                count++;
            } else {
                if (count > 1) {
                    result += count;
                }
                result += s.charAt(i - 1);
                count = 1;
            }
        }

        return result;
    }

    /**
     * Write a function that returns number of valid move of the knight.
     */
    int chessKnight(String cell) {
        int[][] matrix = new int[8][8];
        int count = 0;
        int i1Cell = 7 - (cell.charAt(0) - 97);
        int i2Cell = cell.charAt(1) - 49;
        int X[] = {2, 1, -1, -2, -2, -1, 1, 2};
        int Y[] = {1, 2, 2, 1, -1, -2, -2, -1};

        for (int i = 0; i < matrix.length; i++) {
            for (int j = 0; j < matrix[0].length; j++) {
                matrix[i][j] = 0;
            }
        }

        for (int i = 0; i < 8; i++) {

            int x = i1Cell + X[i];
            int y = i2Cell + Y[i];

            if (x >= 0 && y >= 0 && x < matrix.length && y < matrix[0].length
                    && matrix[x][y] == 0)
                count++;
        }

        return count;
    }

    /**
     * Write a function that returns the maximal number you can obtain
     * by deleting exactly one digit of the given number.
     */
    int deleteDigit(int n) {
        String temp = "";
        int newNumber = 0;
        List<Integer> listNumbers = new ArrayList<Integer>();

        while (n > 0) {
            listNumbers.add(n % 10);
            n /= 10;
        }

        for (int i = 0; i < listNumbers.size(); i++) {
            for (int j = listNumbers.size() - 1; j >= 0; j--) {
                if (j != i) {
                    temp += listNumbers.get(j);
                }
            }

            if (Integer.parseInt(temp) > newNumber) {
                newNumber = Integer.parseInt(temp);
            }
            temp = "";
        }
        return newNumber;
    }
}

File: app/src/at-huybui/java/com/asiantech/intern20summer1/introkotlin/RainbowofClarity.kt
package com.asiantech.intern20summer1.introkotlin

class RainbowofClarity {

    /** 48 **
     * Determine if the given character is a digit or not.
     */
    fun isDigit(symbol: Char): Boolean {
        return symbol.isDigit()
    }

    /** 49 **
     * Given a string, return its encoding defined as follows:
     * First, the string is divided into the least possible number of disjoint substrings
     * consisting of identical characters
     * For example, "aabbbc" is divided into ["aa", "bbb", "c"]
     * Next, each substring with length greater than one is replaced with a concatenation of
     * its length and the repeating character
     * For example, substring "bbb" is replaced by "3b"
     * Finally, all the new strings are concatenated together in the same order and a new string is returned.
     */
    fun lineEncoding(s: String): String {
        var mCharBuf = s[0]
        var mCount = 0;
        var mOutPutString = ""
        s.forEachIndexed { index, c ->
            if (mCharBuf == c) {
                mCount++
            } else {
                if (mCount == 1) {
                    mOutPutString += mCharBuf.toString()
                } else {
                    mOutPutString += mCount.toString() + mCharBuf.toString()
                }
                mCharBuf = c
                mCount = 1
            }
            if (index == s.length - 1) {
                if (mCount == 1) {
                    mOutPutString += mCharBuf.toString()
                } else {
                    mOutPutString += mCount.toString() + mCharBuf.toString()
                }
            }
        }
        return mOutPutString
    }

    /** 50 **
     *  Given a position of a knight on the standard chessboard, find the number of different moves the knight can perform.
     *  The knight can move to a square that is two squares horizontally and one square vertically, or two squares vertically
     *  and one square horizontally away from it. The complete move therefore looks like the letter L.
     *  Check out the image below to see all valid moves for a knight piece that is placed on one of the central squares.
     */
    fun chessKnight(cell: String): Int {
        /** regex
         * 8 ^[c-f][3-6]$
         * 6 ^[c-f][27]|[bg][3-6]$
         * 4 ^[ah][3-6]|[c-f][18]|[bg][27]$
         * 3 ^[ah][27]|[bg][18]$
         * 2 ^[ah][18]$
         */
        return if (cell.matches("^[c-f][3-6]$".toRegex())) {
            8
        } else if (cell.matches("^[c-f][27]|[bg][3-6]$".toRegex())) {
            6
        } else if (cell.matches("^[ah][3-6]|[c-f][18]|[bg][27]$".toRegex())) {
            4
        } else if (cell.matches("^[ah][27]|[bg][18]$".toRegex())) {
            3
        } else if (cell.matches("^[ah][18]$".toRegex())) {
            2
        } else 1
    }


    /** 51 **
     *Given some integer, find the maximal number you can
     * obtain by deleting exactly one digit of the given number.
     */
    fun deleteDigit(n: Int): Int {
        var mMaxValue = 0;  // số lớn nhất xuất ra
        n.toString().forEachIndexed { index, c ->  // quét chuỗi n ->String để bỏ từng phần tử
            val newValue = n.toString().filterIndexed { index2, c2 -> index2 != index }.toInt()
            // hàm trên thực hiện lọc chuỗi n để bỏ phần tử index. sau đó ta gán lại Int ta được new value
            if (newValue > mMaxValue) { // nếu new value lớn hơn max valua => max value = new value
                mMaxValue = newValue
            }
        }
        return mMaxValue
    }

}
